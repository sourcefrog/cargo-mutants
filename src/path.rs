// Copyright 2022 Martin Pool.

//! Utilities for file paths.

use std::convert::TryInto;
use std::fmt;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use camino::{Utf8Path, Utf8PathBuf};
use serde::Serialize;

/// Measures how far above its starting point a path ascends.
///
/// If in following this path you would ever ascend above the starting point,
/// this returns a whole number indicating the number of steps above the
/// starting point.
///
/// This only considers the textual content of the path, and does not look at
/// symlinks or whether the directories actually exist.
pub fn ascent(path: &Utf8Path) -> isize {
    let mut max_ascent: isize = 0;
    let mut ascent = 0;
    for component in path.components().map(|c| c.as_str()) {
        if component == ".." {
            ascent += 1;
        } else if component != "." {
            ascent -= 1;
        }
        if ascent > max_ascent {
            max_ascent = ascent;
        }
    }
    max_ascent
}

/// An extension trait that helps Utf8Path print with forward slashes,
/// even on Windows.
///
/// This makes the output more consistent across platforms and so easier
/// to test.
pub trait Utf8PathSlashes {
    fn to_slash_path(&self) -> String;
}

impl Utf8PathSlashes for Utf8Path {
    fn to_slash_path(&self) -> String {
        self.components()
            .map(|c| c.as_str())
            .filter(|c| !c.is_empty())
            .map(|c| if c == "/" || c == "\\" { "" } else { c })
            .collect::<Vec<_>>()
            .join("/")
    }
}

/// A path relative to the top of the source tree.
#[derive(Debug, PartialEq, Eq, Clone, PartialOrd, Ord, Serialize)]
pub struct TreeRelativePathBuf(Utf8PathBuf);

impl fmt::Display for TreeRelativePathBuf {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = self
            .0
            .components()
            .map(|c| c.as_str())
            .collect::<Vec<_>>()
            .join("/");
        f.write_str(&s)
    }
}

impl TreeRelativePathBuf {
    pub fn new(path: Utf8PathBuf) -> Self {
        assert!(path.is_relative());
        TreeRelativePathBuf(path)
    }

    /// Make an empty tree-relative path, identifying the root.
    #[allow(dead_code)]
    pub fn empty() -> Self {
        TreeRelativePathBuf(Utf8PathBuf::new())
    }

    #[allow(dead_code)]
    pub fn from_absolute(path: &Utf8Path, tree_root: &Utf8Path) -> Self {
        TreeRelativePathBuf(
            path.strip_prefix(tree_root)
                .expect("path is within tree root")
                .to_owned(),
        )
    }

    pub fn within(&self, tree_path: &Utf8Path) -> Utf8PathBuf {
        tree_path.join(&self.0)
    }

    /// Return the tree-relative path of the containing directory.
    ///
    /// Panics if there is no parent, i.e. if self is already the tree root.
    pub fn parent(&self) -> TreeRelativePathBuf {
        self.0
            .parent()
            .expect("TreeRelativePath has no parent")
            .to_owned()
            .into()
    }
}

impl From<&Utf8Path> for TreeRelativePathBuf {
    fn from(path_buf: &Utf8Path) -> Self {
        TreeRelativePathBuf::new(path_buf.to_owned())
    }
}

impl From<Utf8PathBuf> for TreeRelativePathBuf {
    fn from(path_buf: Utf8PathBuf) -> Self {
        TreeRelativePathBuf::new(path_buf)
    }
}

impl From<PathBuf> for TreeRelativePathBuf {
    fn from(path_buf: PathBuf) -> Self {
        TreeRelativePathBuf::new(path_buf.try_into().expect("path must be UTF-8"))
    }
}

impl From<&Path> for TreeRelativePathBuf {
    fn from(path: &Path) -> Self {
        TreeRelativePathBuf::from(path.to_owned())
    }
}

impl FromStr for TreeRelativePathBuf {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(TreeRelativePathBuf::new(s.parse()?))
    }
}

#[cfg(test)]
mod test {
    use camino::{Utf8Path, Utf8PathBuf};

    use super::{ascent, Utf8PathSlashes};

    #[test]
    fn path_slashes_drops_empty_parts() {
        let mut path = Utf8PathBuf::from("/a/b/c/");
        path.push("d/e/f");
        assert_eq!(path.to_slash_path(), "/a/b/c/d/e/f");
    }

    #[test]
    fn path_ascent() {
        assert_eq!(ascent(Utf8Path::new(".")), 0);
        assert_eq!(ascent(Utf8Path::new("..")), 1);
        assert_eq!(ascent(Utf8Path::new("sub/dir")), 0);
        assert_eq!(ascent(Utf8Path::new("sub/dir/../..")), 0);
        assert_eq!(ascent(Utf8Path::new("sub/../sub/./..")), 0);
        assert_eq!(ascent(Utf8Path::new("../back")), 1);
        assert_eq!(ascent(Utf8Path::new("../back/../back")), 1);
        assert_eq!(ascent(Utf8Path::new("../back/../../back/down")), 2);
    }
}
