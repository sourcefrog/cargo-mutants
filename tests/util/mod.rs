// Copyright 2024 Martin Pool

#![allow(dead_code)] // rustc doesn't understand they're used by multiple crates

//! Reusable utilities for cargo-mutants tests.
//!
//! This is available both to integration tests (by `mod util`) and unit tests inside
//! cargo-mutants (as `use crate::test_util`).

use std::borrow::Borrow;
use std::env;
use std::fs::{read_dir, read_to_string, rename};
use std::path::{Path, PathBuf};
use std::time::Duration;

use itertools::Itertools;
use lazy_static::lazy_static;
use tempfile::TempDir;

/// A timeout for a `cargo mutants` invocation from the test suite. Needs to be
/// long enough that even commands that do a lot of work can pass even on slow
/// CI VMs and even on Windows, but short enough that the test does not hang
/// forever.
pub const OUTER_TIMEOUT: Duration = Duration::from_secs(60);

lazy_static! {
    pub static ref MAIN_BINARY: PathBuf = assert_cmd::cargo::cargo_bin("cargo-mutants");
}

pub fn run() -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::new(MAIN_BINARY.as_os_str());
    // Strip any options configured in the environment running these tests,
    // so that they don't cause unexpected behavior in the code under test.
    //
    // For example, without this,
    // `env CARGO_MUTANTS_JOBS=4 cargo mutants`
    //
    // would end up with tests running 4 jobs by default, which would cause
    // the tests to fail.
    //
    // Even more generally than that example, we want the tests to be as hermetic
    // as reasonably possible.
    //
    // Also strip GITHUB_ACTION to avoid automatically emitting github annotations,
    // so that tests are more hermetic and reproducible between local and CI.
    env::vars()
        .map(|(k, _v)| k)
        .filter(|k| {
            k.starts_with("CARGO_MUTANTS_")
                || k == "CLICOLOR_FORCE"
                || k == "NOCOLOR"
                || k == "CARGO_TERM_COLOR"
                || k == "GITHUB_ACTION"
        })
        .for_each(|k| {
            cmd.env_remove(k);
        });
    cmd
}

pub trait CommandInstaExt {
    fn assert_insta(&mut self, snapshot_name: &str);
}

impl CommandInstaExt for assert_cmd::Command {
    fn assert_insta(&mut self, snapshot_name: &str) {
        let output = self.output().expect("command completes");
        assert!(output.status.success());
        insta::assert_snapshot!(snapshot_name, String::from_utf8_lossy(&output.stdout));
        assert_eq!(&String::from_utf8_lossy(&output.stderr), "");
    }
}

// Copy the source for one testdata tree.
pub fn copy_of_testdata(tree_name: &str) -> TempDir {
    assert!(
        !tree_name.contains('/'),
        "testdata tree name {tree_name:?} should be just the directory name"
    );
    let tmp = TempDir::with_prefix(format!("cargo-mutants-testdata-{tree_name}-")).unwrap();
    copy_testdata_to(tree_name, tmp.path());
    tmp
}

pub fn copy_testdata_to<P: AsRef<Path>>(tree_name: &str, dest: P) {
    let dest = dest.as_ref();
    let mut renames = Vec::new();
    cp_r::CopyOptions::new()
        .filter(|path, _stat| {
            Ok(["target", "mutants.out", "mutants.out.old"]
                .iter()
                .all(|p| !path.starts_with(p)))
        })
        .after_entry_copied(|path, _file_type, _stats| {
            if path.ends_with("Cargo_test.toml") || path.ends_with(".cargo_test") {
                renames.push(dest.join(path));
            }
            Ok(())
        })
        .copy_tree(Path::new("testdata").join(tree_name), dest)
        .unwrap();
    for path in &renames {
        let new_name = path
            .file_name()
            .unwrap()
            .to_str()
            .unwrap()
            .replace("_test", "");
        if let Err(err) = rename(path, path.parent().unwrap().join(new_name)) {
            panic!("failed to rename {path:?}: {err:?}")
        }
    }
}

/// Assert that some bytes, when parsed as json, equal a json value.
pub fn assert_bytes_eq_json<J: Borrow<serde_json::Value>>(actual: &[u8], expected: J) {
    // The Borrow is so that you can pass either a value or a reference, for easier
    // calling.
    let actual_json = std::str::from_utf8(actual)
        .expect("bytes are UTF-8")
        .parse::<serde_json::Value>()
        .expect("bytes can be parsed as JSON");
    assert_eq!(&actual_json, expected.borrow());
}

/// Return paths to all testdata trees, in order, excluding leftover git
/// detritus with no Cargo.toml.
pub fn all_testdata_tree_names() -> Vec<String> {
    read_dir("testdata")
        .expect("list testdata")
        .map(|r| r.expect("read testdata dir entry"))
        .filter(|dir_entry| dir_entry.file_type().unwrap().is_dir())
        .filter(|dir_entry| dir_entry.file_name() != "parse_fails")
        .filter(|dir_entry| {
            let dir_path = dir_entry.path();
            dir_path.join("Cargo.toml").exists() || dir_path.join("Cargo_test.toml").exists()
        })
        .map(|dir_entry| {
            dir_entry
                .file_name()
                .into_string()
                .expect("dir name is UTF-8")
        })
        .sorted()
        .collect()
}

pub fn outcome_json(tmp_src_dir: &TempDir) -> serde_json::Value {
    read_to_string(tmp_src_dir.path().join("mutants.out/outcomes.json"))
        .expect("read outcomes.json")
        .parse()
        .expect("parse outcomes.json")
}

pub fn outcome_json_counts(tmp_src_dir: &TempDir) -> serde_json::Value {
    let mut outcomes = outcome_json(tmp_src_dir);
    // We don't want to compare the detailed outcomes
    outcomes.as_object_mut().unwrap().remove("outcomes");
    outcomes
}
