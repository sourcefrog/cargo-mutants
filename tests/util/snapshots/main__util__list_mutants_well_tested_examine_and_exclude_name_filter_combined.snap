---
source: tests/util/mod.rs
expression: "String::from_utf8_lossy(&output.stdout)"
---
src/module/utils/inside_mod.rs:4:13: replace outer::inner::name -> &'static str with ""
src/module/utils/inside_mod.rs:4:13: replace outer::inner::name -> &'static str with "xyzzy"
src/module/utils/sub_utils/subutils_nested_function.rs:2:5: replace has_nested -> u32 with 0
src/module/utils/sub_utils/subutils_nested_function.rs:2:5: replace has_nested -> u32 with 1
src/module/utils/sub_utils/subutils_nested_function.rs:3:9: replace has_nested::inner -> u32 with 0
src/module/utils/sub_utils/subutils_nested_function.rs:3:9: replace has_nested::inner -> u32 with 1
src/module/utils/sub_utils/subutils_nested_function.rs:5:13: replace * with + in has_nested
src/module/utils/sub_utils/subutils_nested_function.rs:5:13: replace * with / in has_nested
